File: infrastructures/ddd/python/src/domain/models/infra/base_model.py
"""
Base class for all Business Model Objects (BMOs).
BMOs contain business logic and invariants.
CRITICAL: BMOs MUST NOT have database/ORM attributes.
They live in /domain/models/ and have Model or BMO suffix.
Separation from persistence layer is MANDATORY.
"""

from abc import ABC
from datetime import datetime
from typing import Optional
import uuid


class BaseModel(ABC):
    """
    Base class for all domain model objects (BMOs).
    
    BMOs are responsible for:
    - Encapsulating business logic
    - Enforcing business invariants
    - Maintaining state consistency
    
    BMOs MUST NOT:
    - Have database/ORM attributes
    - Contain persistence logic
    - Reference entities or data access layer
    """
    
    def __init__(self):
        """Initialize a new BMO with generated ID and timestamps."""
        self._id: str = str(uuid.uuid4())
        self._created_at: datetime = datetime.utcnow()
        self._updated_at: datetime = datetime.utcnow()
    
    @property
    def id(self) -> str:
        """Get the unique identifier of the model."""
        return self._id
    
    @property
    def created_at(self) -> datetime:
        """Get the creation timestamp."""
        return self._created_at
    
    @property
    def updated_at(self) -> datetime:
        """Get the last update timestamp."""
        return self._updated_at
    
    def validate(self) -> None:
        """
        Validates the business invariants of the model.
        Override this method to add specific validation logic.
        
        Raises:
            ValueError: If validation fails
        """
        if not self._id:
            raise ValueError("Model ID cannot be empty")

File: infrastructures/ddd/typescript/src/domain/models/infra/BaseModel.ts
/**
 * Base class for all Business Model Objects (BMOs).
 * This class:
 * - Lives in /domain/models/infra/
 * - Contains common fields and validation logic
 * - MUST NOT have database/ORM attributes
 * - Uses member field naming convention: _variable
 * 
 * Domain models (BMOs):
 * - Contain business logic and invariants
 * - Are pure TypeScript/JavaScript objects
 * - Validate their own state
 * - Are persistence-ignorant
 */
export abstract class BaseModel {
  protected _id: string;
  protected _createdAt: Date;
  protected _updatedAt: Date;

  constructor() {
    this._id = '';
    this._createdAt = new Date();
    this._updatedAt = new Date();
  }

  public get Id(): string {
    return this._id;
  }

  public get CreatedAt(): Date {
    return this._createdAt;
  }

  public get UpdatedAt(): Date {
    return this._updatedAt;
  }

  /**
   * Validates the business invariants of the model.
   * Subclasses should override this method to add their own validation.
   * @throws Error if validation fails
   */
  public Validate(): void {
    if (!this._id) {
      throw new Error('Model ID cannot be empty');
    }
  }
}
